wgengine/monitor/monitor_windows.go

```diff
@@ -7,273 +7,149 @@ package monitor
 import (
 	"context"
 	"errors"
-	"fmt"
-	"runtime"
 	"sync"
-	"syscall"
 	"time"
-	"unsafe"
 
-	"golang.org/x/sys/windows"
-	"tailscale.com/net/interfaces"
+	"golang.zx2c4.com/wireguard/windows/tunnel/winipcfg"
 	"tailscale.com/types/logger"
 )
 
-const (
-	pollIntervalSlow = 15 * time.Second
-	pollIntervalFast = 3 * time.Second
-	pollFastFor      = 30 * time.Second
-)
-
 var (
-	iphlpapi                 = syscall.NewLazyDLL("iphlpapi.dll")
-	notifyAddrChangeProc     = iphlpapi.NewProc("NotifyAddrChange")
-	notifyRouteChangeProc    = iphlpapi.NewProc("NotifyRouteChange")
-	cancelIPChangeNotifyProc = iphlpapi.NewProc("CancelIPChangeNotify")
+	errClosed = errors.New("closed")
 )
 
-const (
-	_STATUS_PENDING = 0x00000103 // 259
-	_STATUS_WAIT_0  = 0
-)
-
-type unspecifiedMessage struct{}
-
-func (unspecifiedMessage) ignore() bool { return false }
-
-type pollStateChangedMessage struct{}
-
-func (pollStateChangedMessage) ignore() bool { return false }
-
-type messageOrError struct {
-	message
-	error
+type eventMessage struct {
+	eventType string
 }
 
-type winMon struct {
-	ctx             context.Context
-	cancel          context.CancelFunc
-	messagec        chan messageOrError
-	logf            logger.Logf
-	pollTicker      *time.Ticker
-	lastState       *interfaces.State
-	closeHandle     windows.Handle // signaled upon close
-	goroutineDoneCh chan struct{}  // closed when awaitIPAndRouteChanges goroutine has stopped
+func (eventMessage) ignore() bool { return false }
 
-	mu            sync.Mutex
-	lastNetChange time.Time
-	inFastPoll    bool // recent net change event made us go into fast polling mode (to detect proxy changes)
+type winMon struct {
+	logf   logger.Logf
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	mu                    sync.Mutex // guards following
+	addressChangeCallback *winipcfg.UnicastAddressChangeCallback
+	routeChangeCallback   *winipcfg.RouteChangeCallback
+	messagec              chan eventMessage
 }
 
 func newOSMon(logf logger.Logf, _ *Mon) (osMon, error) {
-	closeHandle, err := windows.CreateEvent(nil, 1 /* manual reset */, 0 /* unsignaled */, nil /* no name */)
+	m := &winMon{
+		logf: logf,
+	}
+
+	cba, err := winipcfg.RegisterUnicastAddressChangeCallback(m.unicastAddressChanged)
 	if err != nil {
-		return nil, fmt.Errorf("CreateEvent: %w", err)
+		m.logf("winipcfg.RegisterUnicastAddressChangeCallback error: %v", err)
+		return nil, err
 	}
 
-	ctx, cancel := context.WithCancel(context.Background())
-	m := &winMon{
-		logf:            logf,
-		ctx:             ctx,
-		cancel:          cancel,
-		messagec:        make(chan messageOrError, 1),
```

Comment:
I'd bring this back and just create messagec once and not try to track if there's a reader or not. the reader will briefly disappear between calls and we don't want to miss an event in one of those small windows.

Reply:
Ahhhhh, ok, now it makes sense! Originally I understood it wrongly then - I've actually put an effort to prevent _old events_ to pop to the calling code. :) Now everything makes sense.